File: src/support/GBColor.java
/*******************************************************************************
 * Copyright (c) 2002-2013 (c) Devon and Warren Schudy
 * Copyright (c) 2014  Devon and Warren Schudy, Mike Anderson
 *******************************************************************************/
// GBColor.cpp
// Grobots (c) 2002-2006 Devon and Warren Schudy
// Distributed under the GNU General Public License.

package support;

import java.awt.Color;

public class GBColor {
	/**
	 * 
	 */
	// Some standard colors that aren't a named color
	public static final Color darkRed = new Color(139, 0, 0);
	public static final Color darkGreen = new Color(0, 100, 0);
	public static final Color purple = new Color(1, 0, 1);
	public static final Color gold = new Color(0.4f, 0.6f, 0);
	public static final Color shadow = new Color(.2f, .2f, .2f, .5f);

	public static final float kRedWeight = 0.35f;
	public static final float kGreenWeight = 0.45f;
	public static final float kBlueWeight = 0.2f;

	public static final float kMaxTextLightness = 0.7f;

	public static Color getGreyColor(float grey) {
		return new Color(Limit(grey), Limit(grey), Limit(grey));
	}

	/* Not used outside of this class; make private? */
	public static float Lightness(Color c) {
		return c.getRed() * kRedWeight + c.getGreen() * kGreenWeight
				+ c.getBlue() * kBlueWeight / 255;
	}

	public static Color Mix(Color base, float fraction, Color other) {
		return add(multiply(base, fraction), multiply(other, 1.0f - fraction));
	}

	public static float Contrast(Color base, Color other) {
		double r = Math.pow((base.getRed() - other.getRed()) / 255f, 2)
				* kRedWeight;
		double g = Math.pow((base.getGreen() - other.getGreen()) / 255f, 2)
				* kGreenWeight;
		double b = Math.pow((base.getBlue() - other.getBlue()) / 255f, 2)
				* kBlueWeight;
		return (float) Math.sqrt(r + g + b);
	}

	static final float Limit(float val) {
		if (val > 1.0f)
			return 1.0f;
		if (val < 0.0f)
			return 0.0f;
		return val;
	}

	// this function may not return a color with Lightness>minimum if minimum is
	// > 1/3!
	public static Color EnsureContrastWithBlack(Color base, float minimum) {
		float contrast = Contrast(base, Color.BLACK);
		if (contrast >= minimum)
			return base;
		if (contrast == 0.0f)
			return getGreyColor(minimum);
		return multiply(base, minimum / contrast);
	}

	// Returns one of two colors which best contrasts with *this. If the primary
	// color is at least cutoff contrast, secondary is not considered.
	public static Color ChooseContrasting(Color base, Color primary,
			Color secondary, float cutoff) {
		float cp, cs;
		cp = Contrast(base, primary);
		cs = Contrast(base, secondary);
		if (cp >= cutoff)
			return primary;
		else
			return cp >= cs ? primary : secondary;
	}

	public static Color ContrastingTextColor(Color base) {
		if (Lightness(base) < kMaxTextLightness)
			return base;
		else
			return multiply(base, kMaxTextLightness);
	}

	public static Color multiply(Color base, float multiplier) {
		float r = GBMath.clamp(base.getRed() / 255f * multiplier, 0, 1);
		float g = GBMath.clamp(base.getGreen() / 255f * multiplier, 0, 1);
		float b = GBMath.clamp(base.getBlue() / 255f * multiplier, 0, 1);
		return new Color(r, g, b);
	}

	public static Color add(Color base, Color other) {
		int r = GBMath.clamp(base.getRed() + other.getRed(), 0, 255);
		int g = GBMath.clamp(base.getGreen() + other.getGreen(), 0, 255);
		int b = GBMath.clamp(base.getBlue() + other.getBlue(), 0, 255);
		return new Color(r, g, b);
	}

	/**
	 * Returns the hex representation of the given color. Useful for html and
	 * css styling.
	 * 
	 * @param base
	 * @return
	 */
	public static String toHex(Color base) {
		return String.format("#%02x%02x%02x", base.getRed(), base.getGreen(),
				base.getBlue());
	}
}

File: grobots_cpp/support/GBColor.cpp
// GBColor.cpp
// Grobots (c) 2002-2006 Devon and Warren Schudy
// Distributed under the GNU General Public License.

#include "GBColor.h"
#include "GBNumber.h"
#include "GBPlatform.h"
#include <math.h>

// some things are inline, and can be found in GBColor.h
const float kRedWeight = 0.35f;
const float kGreenWeight = 0.45f;
const float kBlueWeight = 0.2f;

const float kMaxTextLightness = 0.7f;


GBColor::GBColor(const float grey)
	: r(Limit(grey)), g(Limit(grey)), b(Limit(grey))
{}

GBColor::GBColor(const float red, const float green, const float blue)
	: r(Limit(red)), g(Limit(green)), b(Limit(blue))
{}

void GBColor::Set(const float red, const float green, const float blue) {
	r = Limit(red);
	g = Limit(green);
	b = Limit(blue);
}

float GBColor::Lightness() const {
	return r * kRedWeight + g * kGreenWeight + b * kBlueWeight;
}

const GBColor GBColor::Mix(const float fraction, const GBColor & other) const {
	return *this * fraction + other * (1.0f - fraction);
}

const float GBColor::Contrast(const GBColor & other) const {
	return sqrt((r - other.r) * (r - other.r) * kRedWeight +
			(g - other.g) * (g - other.g) * kGreenWeight +
			(b - other.b) * (b - other.b) * kBlueWeight);
	//alternate, maybe better
	return fabs(r - other.r) * kRedWeight +
			fabs(g - other.g) * kGreenWeight +
			fabs(b - other.b) * kBlueWeight;
}


float GBColor::Limit(float val) {
	if ( val > 1.0f ) return 1.0f;
	if ( val < 0.0f ) return 0.0f;
	return val;
}

// this function may not return a color with Lightness>minimum if minimum is > 1/3!
const GBColor GBColor::EnsureContrastWithBlack(const float minimum) const {
	float contrast = Contrast(GBColor::black);
	if ( contrast >= minimum )
		return *this;
	if ( contrast == 0.0f )
		return GBColor(minimum);
	return *this * (minimum / contrast);
}

// Returns one of two colors which best contrasts with *this. If the primary color
//  is at least cutoff contrast, secondary is not considered.
const GBColor GBColor::ChooseContrasting(const GBColor & primary, const GBColor & secondary,
		const float cutoff) const {
	float cp, cs;
	cp = Contrast(primary);
	cs = Contrast(secondary);
	if ( cp >= cutoff )
		return primary;
	else
		return cp >= cs ? primary : secondary;
}

const GBColor GBColor::ContrastingTextColor() const {
	if ( Lightness() < kMaxTextLightness )
		return *this;
	else
		return *this * kMaxTextLightness;
}

const GBColor GBColor::operator *(float multiplier) const {
	return GBColor(r * multiplier, g * multiplier, b * multiplier);
}

const GBColor GBColor::operator /(float divisor) const {
	if ( ! divisor ) throw GBDivideByZeroError();
	return GBColor(r / divisor, g / divisor, b / divisor);
}

const GBColor GBColor::operator +(const GBColor & other) const {
	return GBColor(r + other.r, g + other.g, b + other.b);
}

const GBColor GBColor::red(1, 0, 0);
const GBColor GBColor::green(0, 1, 0);
const GBColor GBColor::blue(0, 0, 1);
const GBColor GBColor::cyan(0, 1, 1);
const GBColor GBColor::magenta(1, 0, 1);
const GBColor GBColor::yellow(1, 1, 0);
const GBColor GBColor::black(0);
const GBColor GBColor::white(1);
const GBColor GBColor::gray(0.5f);
const GBColor GBColor::lightGray(0.8f);
const GBColor GBColor::darkGray(0.2f);
const GBColor GBColor::purple(0.6f, 0, 0.8f);
const GBColor GBColor::darkGreen(0, 0.5f, 0);
const GBColor GBColor::darkRed(0.7f, 0, 0);

